toothpick-runtime/src/main/java/toothpick/Configuration.java

```diff
@@ -4,100 +4,182 @@
-      // We need a LIFO structure here, but stack is thread safe and we use thread local,
-      // so this property is overkill and LinkedHashSet is faster on retrieval.
-      private ThreadLocal<LinkedHashSet<Class>> cycleDetectionStack = new ThreadLocal<LinkedHashSet<Class>>() {
-        @Override
-        protected LinkedHashSet<Class> initialValue() {
-          return new LinkedHashSet<>();
-        }
-      };
+    return new DevelopmentConfiguration();
+  }
 
-      @Override
-      void checkIllegalBinding(Binding binding) {
-        Class<?> clazz;
-        switch (binding.getMode()) {
-          case SIMPLE:
-            clazz = binding.getKey();
-            break;
-          case CLASS:
-            clazz = binding.getImplementationClass();
-            break;
-          case PROVIDER_CLASS:
-            clazz = binding.getProviderClass();
-            break;
-          default:
-            return;
-        }
+  /**
+   * Performs no runtime checks. This configuration
+   * is fast but less than {@link #reflectionFree()}.
+   * It can be used in production.
+   * This configuration uses reflection.
+   *
+   * @return a production configuration.
+   */
+  public static Configuration production() {
+    return new ProductionConfiguration();
+  }
 
-        for (Annotation annotation : clazz.getAnnotations()) {
-          if (annotation.annotationType().isAnnotationPresent(javax.inject.Scope.class)) {
-            throw new IllegalBindingException(format("Class %s cannot be bound. It has an scope annotation", clazz.getName()));
-          }
-        }
-      }
+  /**
+   * Performs no runtime checks. This configuration
+   * is the fastest but will need additional setup
+   * of the annotation processors.
+   * It can be used in production.
+   * This configuration doesn't use reflection.
+   *
+   * @return an optimized reflection free configuration.
+   */
+  public static Configuration reflectionFree() {
+    return new ReflectionFreeConfiguration();
+  }
 
-      @Override
-      void checkCyclesStart(Class clazz) {
-        if (cycleDetectionStack.get().contains(clazz)) {
-          throw new CyclicDependencyException(new ArrayList<>(cycleDetectionStack.get()), clazz);
-        }
+  private static class BaseConfiguration extends Configuration {
 
-        cycleDetectionStack.get().add(clazz);
-      }
+    @Override
+    public void checkIllegalBinding(Binding binding) {
+    }
+
+    @Override
+    public void checkCyclesStart(Class clazz) {
+    }
+
+    @Override
+    public void checkCyclesEnd(Class clazz) {
+    }
+
+    @Override
+    public <T> Factory<T> getFactory(Class<T> clazz) {
+      return FactoryRegistryLocator.getFactoryUsingReflection(clazz);
+    }
+
+    @Override
+    public <T> MemberInjector<T> getMemberInjector(Class<T> clazz) {
+      return MemberInjectorRegistryLocator.getMemberInjectorUsingReflection(clazz);
+    }
+  }
 
+  private static class DevelopmentConfiguration extends Configuration {
```

Comment:
This can extend BaseConfiguration. So getFactory and getMemberInjector are already defined.